calyx-py/calyx/queues.py

```diff
@@ -599,8 +599,98 @@ def increment_hot(self):
         self.hot = 0 if self.hot == (self.n - 1) else self.hot + 1
 
     def pop(self, *_) -> Optional[int]:
-        """Pops the PIFO by popping some internal FIFO.
-        Updates `hot` to be one more than the index of the internal FIFO that
+        """Pops the PIFO by popping some internal subqueue.
+        Updates `hot` to be one more than the index of the internal subqueue that
+        we did pop.
+        """
+        if self.pifo_len == 0:
+            raise QueueError("Cannot pop from empty PIFO.")
+
+        while True:
+            try:
+                val = self.data[self.hot].pop()
+                if val is not None:
+                    self.increment_hot()
+                    self.pifo_len -= 1
+                    return val
+                self.increment_hot()
+            except QueueError:
+                self.increment_hot()
+
+    def peek(self, *_) -> Optional[int]:
+        """Peeks into the PIFO. Does not affect what `hot` is."""
+        if self.pifo_len == 0:
+            raise QueueError("Cannot peek into empty PIFO.")
+
+        original_hot = self.hot
+        while True:
+            try:
+                val = self.data[self.hot].peek()
+                if val is not None:
+                    self.hot = original_hot
+                    return val
+                self.increment_hot()
+            except QueueError:
+                self.increment_hot()
+
+    def __len__(self) -> int:
+        return self.pifo_len
+
+@dataclass
+class RRQueue2:
```

Comment:
Why do we need `RRQueue` and `RRQueue2`?

Reply:
I had meant to merge the two versions; this has been resolved so that RRQueue now takes in a `subqueues` parameter instead of assuming subqueues that are FIFOs.

Reply:
Super!